lang/cs/Org.Apache.REEF.Client/AzureBatch/Service/ContainerRegistryProvider.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using Microsoft.Azure.Batch;
+using Org.Apache.REEF.Client.AzureBatch.Parameters;
+using Org.Apache.REEF.Tang.Annotations;
+
+
+namespace Org.Apache.REEF.Client.DotNet.AzureBatch
+{
+    public class ContainerRegistryProvider
+    {
+        public string ContainerRegistryServer { get; private set; }
+        public string ContainerRegistryUsername { get; private set; }
+        public string ContainerRegistryPassword { get; private set; }
+        public string ContainerImageName { get; private set; }
+
+        [Inject]
+        public ContainerRegistryProvider(
+            [Parameter(typeof(ContainerRegistryServer))] string containerRegistryServer,
+            [Parameter(typeof(ContainerRegistryUsername))] string containerRegistryUsername,
+            [Parameter(typeof(ContainerRegistryPassword))] string containerRegistryPassword,
+            [Parameter(typeof(ContainerImageName))] string containerImageName
+      )
+        {
+            this.ContainerRegistryServer = containerRegistryServer;
+            this.ContainerRegistryUsername = containerRegistryUsername;
+            this.ContainerRegistryPassword = containerRegistryPassword;
+            this.ContainerImageName = containerImageName;
+        }
+
+        public bool IsValid()
+        {
+            return !string.IsNullOrEmpty(this.ContainerRegistryServer);
```

Comment:
that can be a public readonly property precomputed in the constructor. so I would write something like this:
```C#
public sealed class ContainerRegistryProvider
{
    private readonly string _containerRegistryServer;
    private readonly string _containerRegistryUsername;
    private readonly string _containerRegistryPassword;

    public bool IsValid { get; }

    [Inject]
    public ContainerRegistryProvider(
        [Parameter(typeof(ContainerRegistryServer))] string containerRegistryServer,
        [Parameter(typeof(ContainerRegistryUsername))] string containerRegistryUsername,
        [Parameter(typeof(ContainerRegistryPassword))] string containerRegistryPassword)
    {
        _containerRegistryServer = containerRegistryServer;
        _containerRegistryUsername = containerRegistryUsername;
        _containerRegistryPassword = containerRegistryPassword;
        IsValid = !string.IsNullOrEmpty(containerRegistryServer);
    }

    public ContainerRegistry GetContainerRegistry()
    {
        return IsValid ? new ContainerRegistry(
            userName: _containerRegistryUsername,
            registryServer: _containerRegistryServer,
            password: _containerRegistryPassword) : null;
    }
}
```

Reply:
I see two suggestions in here:
1. Change IsValid to be a property than a method: Given how I worded this .. using a verb (IsValid rather than Valid) .. I feel IsValid must be  method and not a property. If the concern is that the use of a method is slower (since the check is done every time if inside a method as opposed to properties that can be used directly once they are set inside the constructors: C# does optimize code away and I very much expect this to be optimized away as well (although I cannot say that authoritatively, I am not an expert).
2. The proposal to use member variables instead of properties. Using get/set properties causes C# to internally define the member variables (which are hidden to the developer). The use of properties is just a C# best practice as it leaves it open to extension later if needed. I did it just because it is a best practice. Nothing else to it.

Won't fixing for now. If you still disagree (perhaps this is breaking away from the convention for other C# code that is somewhat irksome), please respond back and I will make the changes as proposed.


---
In reply to: [203904916](https://github.com/apache/reef/pull/1478#discussion_r203904916) [](ancestors = 203904916)